crates/zksync/compiler/src/zksolc/mod.rs

```diff
@@ -33,67 +36,113 @@ pub struct DualCompiledContract {
+                        name: contract_name,
+                        zk_bytecode_hash: packed_bytecode.bytecode_hash(),
+                        zk_deployed_bytecode: packed_bytecode.bytecode(),
+                        zk_factory_deps: packed_bytecode.factory_deps(),
+                        evm_bytecode_hash: keccak256(solc_deployed_bytecode),
+                        evm_bytecode: solc_bytecode.to_vec(),
+                        evm_deployed_bytecode: solc_deployed_bytecode.to_vec(),
+                    });
+                }
             }
         }
+
+        Self { contracts: dual_compiled_contracts }
     }
 
-    dual_compiled_contracts
-}
+    /// Finds a contract matching the ZK deployed bytecode
+    pub fn find_by_zk_deployed_bytecode(&self, bytecode: &[u8]) -> Option<&DualCompiledContract> {
+        self.contracts.iter().find(|contract| bytecode.starts_with(&contract.zk_deployed_bytecode))
+    }
 
-/// Implements methods to look for contracts
-pub trait FindContract {
     /// Finds a contract matching the EVM bytecode
-    fn find_evm_bytecode(&self, bytecode: &[u8]) -> Option<&DualCompiledContract>;
+    pub fn find_by_evm_bytecode(&self, bytecode: &[u8]) -> Option<&DualCompiledContract> {
+        self.contracts.iter().find(|contract| bytecode.starts_with(&contract.evm_bytecode))
+    }
 
     /// Finds a contract matching the ZK bytecode hash
-    fn find_zk_bytecode_hash(&self, code_hash: H256) -> Option<&DualCompiledContract>;
-}
+    pub fn find_by_zk_bytecode_hash(&self, code_hash: H256) -> Option<&DualCompiledContract> {
+        self.contracts.iter().find(|contract| code_hash == contract.zk_bytecode_hash)
+    }
+
+    /// Finds a contract own and nested factory deps
+    pub fn fetch_all_factory_deps<'s>(&'s self, root: &'s DualCompiledContract) -> HashSet<&[u8]> {
+        let mut visited = HashSet::new();
+        let mut queue = VecDeque::new();
+
+        for dep in &root.zk_factory_deps {
+            queue.push_back(dep.as_slice());
+        }
+
+        while let Some(dep) = queue.pop_front() {
+            //try to insert in the list of visited, if it's already present, skip
+            if visited.insert(dep) {
+                if let Some(contract) = self.find_by_zk_deployed_bytecode(dep) {
+                    debug!(
+                        name = contract.name,
+                        deps = contract.zk_factory_deps.len(),
+                        "new factory depdendency"
+                    );
+
+                    for nested_dep in &contract.zk_factory_deps {
+                        //check that the nested dependency is inserted
+                        if !visited.contains(nested_dep.as_slice()) {
+                            //if not, add it to queue for processing
+                            queue.push_back(nested_dep.as_slice());
+                        }
+                    }
+                }
+            }
+        }
+
+        visited
```

Comment:
Partially agree and partially disagree: having a `HashSet` allows us to avoid having to deduplicate, or at least ensure this at the type level, so even appending items (like in create with `--factory-deps`) will be guaranteed to skip duplicates, and we also get faster lookups (hence why it's used internally).
On the other hand, we do mainly use `Vec<Vec<u8>>` when transmitting this information, so yeah when we actually use this set we will always do the conversion.

IMHO returning more convenient type is not worth as much as the information lost.